### ISSUE TITLE
CKV_AWS_45 false positives

### ISSUE THREAD

**Issue by solita-mao** [bridgecrewio/checkov#7310]
**Describe the issue**
CKV_AWS_45 "Lambda function’s environment variables expose secrets" reports findings even there are no secrets in Lambda environment. 

**Examples**
Example A: (CDK Bucket Deployment)
`{
  "Environment": {
    "Variables": {
      "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem"
    }
  }
}`

Debug logs:
`2025-09-23 09:43:54,798 [MainThread  ] [DEBUG]  should_run_check CKV_AWS_45: True
2025-09-23 09:43:54,798 [MainThread  ] [DEBUG]  Running check: Ensure no hard-coded secrets exist in Lambda environment on file /tf/demoStack0668BA2D.template.json
2025-09-23 09:43:54,798 [MainThread  ] [DEBUG]  File /tf/demoStack0668BA2D.template.json, resource "AWS::Lambda::Function.CustomCDKBucketDeployment8693BB64968944B69ACFB0CB9EB8756C81C01536" check "Ensure no hard-coded secrets exist in Lambda environment" Result: {'result': <CheckResult.FAILED: 'FAILED'>, 'evaluated_keys': ['Properties/Environment/Variables/__file__']}`

EXAMPLE B: (Secrets Manager used for secrets handling)
`{
  "Environment": {
    "Variables": {
      "DB_USER": "demo",
      "DB_HOST": {
        "Fn::GetAtt": ["demoDatabaseClusterA918C4BA", "Endpoint.Address"]
      },
      "DB_NAME": "demo",
      "DB_SECRET_NAME": {
        "Fn::Join": [
          "-",
          [
            {
              "Fn::Select": [
                0,
                {
                  "Fn::Split": [
                    "-",
                    {
                      "Fn::Select": [
                        6,
                        {
                          "Fn::Split": [
                            ":",
                            {
                              "Ref": "demoDatabaseSecretB0AB457E"
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            },
            {
              "Fn::Select": [
                1,
                {
                  "Fn::Split": [
                    "-",
                    {
                      "Fn::Select": [
                        6,
                        {
                          "Fn::Split": [
                            ":",
                            {
                              "Ref": "demoDatabaseSecretB0AB457E"
                            }
                          ]
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        ]
      },
      "DB_PORT": {
        "Fn::GetAtt": ["demoDatabaseClusterA918C4BA", "Endpoint.Port"]
      }
    }
  }
}
`
Debug logs:
`2025-09-23 09:48:51,214 [MainThread  ] [DEBUG]  should_run_check CKV_AWS_45: True
2025-09-23 09:48:51,214 [MainThread  ] [DEBUG]  Running check: Ensure no hard-coded secrets exist in Lambda environment on file /tf/demoStackB73D57F6.template.json
2025-09-23 09:48:51,214 [MainThread  ] [DEBUG]  File /tf/demoStackB73D57F6.template.json, resource "AWS::Lambda::Function.demoPostgreSQLIamUserD2E888BF" check "Ensure no hard-coded secrets exist in Lambda environment" Result: {'result': <CheckResult.FAILED: 'FAILED'>, 'evaluated_keys': ['Properties/Environment/Variables/__file__']}
`

**Version (please complete the following information):**
 - Checkov Version 3.2.439 onwards

**Additional context**

**Comment by srija1102** [comment]
✅ Verification Update

I retested this issue to confirm whether the false-positive behavior in CKV_AWS_45 (“Lambda function’s environment variables expose secrets”) still persists in newer Checkov releases.

To reproduce the original conditions, I recreated both examples provided in the issue description:

Example A: A Lambda environment containing a non-sensitive system variable (AWS_CA_BUNDLE path).

Example B: A Lambda environment using CloudFormation intrinsic functions such as Fn::GetAtt, Fn::Join, and Ref for secret management via AWS Secrets Manager.

Both templates were scanned using Checkov v3.2.484 with the following commands:

checkov -f lambda_no_secret.json --check CKV_AWS_45 -o cli
checkov -f lambda_with_intrinsics.json --check CKV_AWS_45 -o cli


The scans completed successfully, and both checks passed without any false positives. There were no parsing errors, and the previous debug signature
evaluated_keys: ['Properties/Environment/Variables/__file__']
did not appear, indicating that the parsing and evaluation logic has been corrected.

🧩 Conclusion

The CKV_AWS_45 false-positive issue originally reported in v3.2.439 no longer occurs in v3.2.484 and later. Both the CDK Bucket Deployment and intrinsic function scenarios now pass as expected.

✅ Conclusion: The bug has been resolved, and this issue can be considered closed.

**Comment by solita-mao** [comment]
Still getting false positives with Checkov v3.2.484.
Could the reason be e.g. these kind of lines (no secrets), where it finds a certain string:

"Ref": "demoDatabaseSecretB0AD748E"
or
"S3Key": "1fcecf40b0372b407832c51b1e2ac8b56e63677f19073da1b976feddc51f88d6.zip"

**Comment by solita-mao** [comment]
Not using Checkov anymore.

**Comment by nwmonroe11** [comment]
**Root cause for this issue: the flagged value is checkov's own `__file__`
key, not a template variable.**

The debug log in the original report already shows it:
`evaluated_keys: ['Properties/Environment/Variables/__file__']`.

The CloudFormation parser injects a synthetic `__file__` entry into every
parsed `Environment/Variables` map, holding the path checkov was given.
`LambdaEnvironmentCredentials.scan_resource_conf` iterates that map and
scans every non-dict value, so the file path is scanned as though the
template declared it as an environment variable.

That explains two things the existing reports find puzzling:

1. **Resources whose variables are all intrinsics still fail.** The check
   skips dict values with the comment "if it is a resolved intrinsic function
   like `!Ref: xyz`, then it can't be a secret" — but `__file__` is a plain
   string, so it is scanned regardless.
2. **The result depends on how the file is addressed.** An AWS secret access
   key is 40 characters of `[A-Za-z0-9/+=]`. Whether the path contains a
   maximal run of exactly that length decides the outcome, so the same file
   passes or fails depending on the checkout location.

On a GitHub Actions runner the workspace path is
`/home/runner/work/<repo>/<repo>/...`, and for one repository the substring
`/home/runner/work/<repo>/<repo>/backend/<repo>` was exactly 40 such
characters. Every real environment variable scanned clean; the only match
was the path. Instrumenting `get_secrets_from_string`:

```
SCANNED 'example.org,example.com' -> []
SCANNED 'false'                   -> []
SCANNED '30'                      -> []
SCANNED '72'                       -> []
SCANNED '/home/runner/work/<repo>/<repo>/backend/<repo>-backend.yaml'
     -> ['/home/runner/work/<repo>/<repo>/backend/<repo>']
```

### Minimal reproduction

The template's only environment variable is a `!Ref`, which the check is
written to skip:

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Resources:
  Table:
    Type: AWS::DynamoDB::Table
    Properties:
      BillingMode: PAY_PER_REQUEST
      AttributeDefinitions: [{AttributeName: id, AttributeType: S}]
      KeySchema: [{AttributeName: id, KeyType: HASH}]
  Fn:
    Type: AWS::Lambda::Function
    Properties:
      Runtime: python3.12
      Handler: index.handler
      Role: !GetAtt Role.Arn
      Code: {ZipFile: "def handler(e,c): return 1\n"}
      Environment:
        Variables:
          TABLE_NAME: !Ref Table
  Role:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Version: '2012-10-17'
        Statement: [{Effect: Allow, Principal: {Service: lambda.amazonaws.com}, Action: sts:AssumeRole}]
```

Place it so that the absolute path contains a maximal `[A-Za-z0-9/+=]` run of
exactly 40 characters. `<base>/x-home/runner/work/devi/deviz/backend/devi-backend.yaml`
works, because `home/runner/work/devi/deviz/backend/devi` is 40 characters
and is bounded by the two hyphens:

```console
$ checkov -f "$PWD/devi-backend.yaml" --framework cloudformation --compact
CKV_AWS_45  AWS::Lambda::Function.Fn   FAILED

$ checkov -f devi-backend.yaml --framework cloudformation --compact
CKV_AWS_45  AWS::Lambda::Function.Fn   PASSED
```

Same file, same working directory, same process — only the argument form
differs.

### Expected vs actual

**Expected:** parser bookkeeping is not scanned as user content. `__file__`
is not something a template author wrote.
**Actual:** it is scanned, and a scan result depends on where the repository
happens to sit on disk.

### Suggested fix

Skip keys the parser injects, in `scan_resource_conf`:

```python
for var_name, value in variables.items():
    if var_name.startswith("__") and var_name.endswith("__"):
        continue
```

A more general fix would exclude parser metadata before any check sees the
resource configuration, which would cover other SECRETS-category checks that
iterate mappings the same way.

### Workaround

Pass a repository-relative path rather than an absolute one, so the
filesystem layout does not enter the matcher. No suppression is needed and
the check keeps evaluating every real variable.

### Environment

checkov 3.3.16 · CloudFormation framework · Python 3.12 · reproduced on both
`ubuntu-latest` (GitHub Actions) and macOS 15 (arm64).
